2-Draft-Accepted/RFCxxxx-Secrets-Management.md

```diff
@@ -0,0 +1,150 @@
+
+Secrets are stored securely in a local vault.
+The local vault is expected to only allow access to the user who owns that
+vault.
+Secrets required to access remote vaults are stored in the local vault and used by the secrets management
+cmdlets to retrieve remote secrets.
+
+`User Context` --> `Local Vault` --> `SecretsVaultExtension` --> `Remote Vault`
+
+## User Experience
+
+Registering and using remote secrets:
+
+```powershell
+# Install the Az.KeyVault module to be used in this example.
+# This module doesn't conform to the extensions model so would need to be
+# updated to actually work.
+Install-Module Az.KeyVault
+
+# In this example, we explicitly register this extension
+Register-SecretsVaultExtension -Name Azure -Cmdlet Get-AzKeyVaultSecret `
+  -Module Az.KeyVault -Version 1.0.0
+
+# When using remote vaults, expectation is that secrets are
+# stored using their tooling, so this module is only for retrieving secrets
+# from remote vaults
+$azDevOpsToken = Get-Secret -Name AzDevOps
+Invoke-RestMethod https://dev.azure.com/… -Credential $azDevOpsToken -Authentication Basic
+```
+
+Registering and using local secret:
+
+```powershell
+# For local vault, we can register custom secrets
+# In this example, we store a PSCredential object
+Register-Secret -Secret $cred -Name MyCreds
+New-PSSession -ComputerName myServer -Credential (Get-Secret -Name MyCreds)
+```
+
+## Specification
+
+### Secrets Vault
+
+There are two types of vaults: local and remote.
+A local vault, by default, is already created and named `Default`.
+`Default` on Windows is [CredMan](https://docs.microsoft.com/en-us/windows/desktop/SecAuthN/credentials-management).
+`Default` on macOS is [KeyChain](https://developer.apple.com/documentation/security/keychain_services).
+
+>[!NOTE]
+>KeyChain support is unlikely to be available in the first release of this feature.
+
+`Default` on Linux is [Hashicorp Vault](https://www.vaultproject.io/).
+
+>[!NOTE]
+>Linux has many options for local credential management, additional support may be
+>added in the future or via community provided extensions.
+
+### Credential Vault Extensions
+
+Vault extensions whether they are local or remote are provided as modules.
+
+Extensions are modules that provide a `Get-Secret` cmdlet that accepts a
+`-Credential` or `-AccessToken` parameter for authentication.
+Expectation is that each module will have their own cmdlet and not literally
+named `Get-Secret`.
+
+Extensions register themselves by calling `Register-SecretsVaultExtension`
+passing the name of the relevant cmdlet to `-Cmdlet`, the module name to `-Module`,
+and the module version to `-ModuleVersion`.
+This can be done as part of importing the module or explicitly by the user.
+
+`Register-SecretsVaultExtension` has a `-Default` switch to register the
+default local vault (if applicable, see above) if it has been unregistered.
+
+`Register-SecretsVaultExtension` takes a `-AccessToken` or `-Credential` parameter
```

Comment:
I don't like -AccessToken or -Credential because this makes incorrect assumptions as to what the provider needs. Instead of these (or perhaps in addition to), it should allow you to specify a PSObject or Hashtable called -AuthenticationParameters to make it provider agnostic.

For example, AWS auth using roles requires AccessKey, SecretKey and SessionToken. I could hack around it and provide the SessionToken to -AccessToken and Access/SecretKey to -Credential, but that's dirty.

And what if my provider requires Mutual TLS? Where would I specify my Certificate?

Reply:
I'm adding support for a hashtable that will get splatted, was trying to come up with a general case as a hashtable doesn't provide tab completion.  However, I expect most users to cut & paste registration or have it handled automatically via a module so perhaps we can simplify it for now to just a hashtable and add specific params as standards emerge.